pkg/util/vsphere/watcher/watcher.go

```diff
@@ -0,0 +1,676 @@
+		}
+	}(fromContext(ctx).rem)
+
+	go func() {
+		defer w.Close()
+
+		if err := w.pc.WaitForUpdatesEx(
+			ctx,
+			&property.WaitOptions{},
+			func(ou []vimtypes.ObjectUpdate) bool {
+				return w.onUpdate(ctx, ou)
+			}); err != nil {
+
+			w.setErr(err)
+		}
+	}()
+
+	return w, nil
+}
+
+const (
+	virtualMachineType               = "VirtualMachine"
+	configPropPath                   = "config"
+	extraConfigPropPath              = configPropPath + ".extraConfig"
+	extraConfigNamespaceNameKey      = "vmservice.namespacedName"
+	extraConfigNamespaceNamePropPath = extraConfigPropPath + `["` + extraConfigNamespaceNameKey + `"]`
+)
+
+type objUpdate struct {
+	kind    vimtypes.ObjectUpdateKind
+	changes []vimtypes.PropertyChange
+}
+
+func (w *Watcher) onUpdate(
+	ctx context.Context,
+	ou []vimtypes.ObjectUpdate) bool {
+
+	logger := logr.FromContextOrDiscard(ctx)
+	logger.V(4).Info("onUpdate", "objectUpdates", ou)
+
+	updates := map[moRef]objUpdate{}
+
+	for i := range ou {
+		oui := ou[i]
+		if oui.Kind != vimtypes.ObjectUpdateKindLeave {
+			if v, ok := updates[oui.Obj]; !ok {
+				updates[oui.Obj] = objUpdate{
+					kind:    oui.Kind,
+					changes: oui.ChangeSet,
+				}
+			} else {
+				v.changes = append(v.changes, oui.ChangeSet...)
+				updates[oui.Obj] = v
+			}
+		}
+	}
+
+	for obj, update := range updates {
+		if err := w.onObject(
+			ctx,
+			obj,
+			update); err != nil {
+
+			w.setErr(err)
+			return true
+		}
+	}
+
+	return false
+}
+
+func (w *Watcher) onObject(
+	ctx context.Context,
+	obj moRef,
+	update objUpdate) error {
+
+	logger := logr.FromContextOrDiscard(ctx).
+		V(4).
+		WithName("onObject").
+		WithValues("obj", obj)
+
+	var (
+		namespace string
+		name      string
+		verified  bool
+	)
+
+	// This update will be skipped if after removing all of the changes for
+	// the ignoredExtraConfigKeys there is nothing left.
+	var ignoredChanges int
+	for i := range update.changes {
+		c := update.changes[i]
+		ignore := false
+
+		if c.Name == extraConfigPropPath {
+			if aov, ok := c.Val.(vimtypes.ArrayOfOptionValue); ok {
+				var ignoredOptVals int
+				for j := range aov.OptionValue {
+					if ov := aov.OptionValue[j].GetOptionValue(); ov != nil {
+
+						// Get the namespace and name of the VM from the changes
+						// if they are present there.
+						if ov.Key == extraConfigNamespacedNameKey {
+							if namespace == "" || name == "" {
+								if s, ok := ov.Value.(string); ok {
+									namespace, name = namespacedNameFromString(s)
+								}
+							}
+						}
+
+						// Note if the key is ignored.
+						if _, ok := w.ignoredExtraConfigKeys[ov.Key]; ok {
+							ignoredOptVals++
+						}
+					}
+				}
+
+				// The change to extraConfig is ignored if the number of ignored
+				// option values is the same as the number of total option vals.
+				ignore = ignoredOptVals == len(aov.OptionValue)
+			}
+		}
+		if ignore {
+			ignoredChanges++
+		}
+	}
+	if ignoredChanges == len(update.changes) {
+		return nil
+	}
```

Comment:
Instead of counting all the ignored EC keys, we could also look for a non-ignored EC key to break the loop earlier if it's found and namespace & name are set. Something like:

```suggestion
func (w *Watcher) onObject(
	...

	// This update will be skipped if after removing all of the changes for
	// the ignoredExtraConfigKeys there is nothing left.
	var ignoredChanges int
	for i := range update.changes {
		c := update.changes[i]
		ignore := false

		if c.Name == extraConfigPropPath {
			if aov, ok := c.Val.(vimtypes.ArrayOfOptionValue); ok {
				ignore, namespace, name = checkEC(aov, w.ignoredExtraConfigKeys)
			}
		}
		if ignore {
			ignoredChanges++
		}
	}
	if ignoredChanges == len(update.changes) {
		return nil
	}
	
	...
}

func checkEC(
	aov vimtypes.ArrayOfOptionValue,
	ignoredKeys map[string]struct{}) (ignore bool, namespace, name string) {

	hasNonIgnoredECKey := false

	for j := range aov.OptionValue {
		if ov := aov.OptionValue[j].GetOptionValue(); ov != nil {

			// Get the namespace and name of the VM from the changes
			// if they are present there.
			if ov.Key == extraConfigNamespacedNameKey {
				if namespace == "" || name == "" {
					if s, ok := ov.Value.(string); ok {
						namespace, name = namespacedNameFromString(s)
					}
				}
			}

			// Note if the key cannot be ignored.
			if _, ok := ignoredKeys[ov.Key]; !ok {
				hasNonIgnoredECKey = true
			}

			if hasNonIgnoredECKey && (namespace != "" && name != "") {
				break
			}
		}
	}

	return !hasNonIgnoredECKey, namespace, name
}
```